hedera-node/hedera-app/src/main/java/com/hedera/node/app/quiescence/QuiescenceController.java

```diff
@@ -0,0 +1,215 @@
+        signedTransactionParser = createParser(SignedTransaction.PROTOBUF);
+        nextTct = new AtomicReference<>();
+        pipelineTransactionCount = new AtomicLong(0);
+    }
+
+    /**
+     * Notifies the controller that a block has been fully signed.
+     *
+     * @param block the fully signed block
+     */
+    public void fullySignedBlock(@NonNull final Block block) {
+        if (!config.enabled()) {
+            return;
+        }
+        // FOR EXECUTION REVIEWERS:
+        // Can we avoid parsing transactions here?
+        final long transactionCount = block.items().stream()
+                .filter(BlockItem::hasSignedTransaction)
+                .map(BlockItem::signedTransaction)
+                .filter(Objects::nonNull)
+                .filter(this::isRelevantTransaction)
+                .count();
+        final long updatedValue = pipelineTransactionCount.addAndGet(-transactionCount);
+        if (updatedValue < 0) {
+            logger.error("Quiescence transaction count overflow, turning off quiescence");
+            disableQuiescence();
+        }
+        // FOR EXECUTION REVIEWERS:
+        // Is this the correct way to get consensus time?
+        final Optional<Timestamp> maxConsensusTime = block.items().stream()
+                .filter(BlockItem::hasStateChanges)
+                .map(BlockItem::stateChanges)
+                .filter(Objects::nonNull)
+                .map(StateChanges::consensusTimestamp)
+                .max(HapiUtils.TIMESTAMP_COMPARATOR);
+        if (maxConsensusTime.isEmpty()) {
+            return;
+        }
+        final Instant maxConsensusInstant = HapiUtils.asInstant(maxConsensusTime.get());
+        nextTct.accumulateAndGet(maxConsensusInstant, QuiescenceController::tctUpdate);
+    }
+
+    /**
+     * Notifies the controller that an event has been and will be handled soon (if it doesn't become stale).
+     *
+     * @param event the event that will be handled
+     */
+    public void onPreHandle(@NonNull final Event event) {
+        if (!config.enabled()) {
+            return;
+        }
+        // FOR EXECUTION REVIEWERS:
+        // Should we be parsing the transactions here? I would assume they have already been parsed
+        pipelineTransactionCount.addAndGet(countRelevantTransactions(event));
+    }
+
+    /**
+     * Notifies the controller that an event has become stale and will not be handled.
+     *
+     * @param event the event that has become stale
+     */
+    public void staleEvent(@NonNull final Event event) {
+        if (!config.enabled()) {
+            return;
+        }
+        pipelineTransactionCount.addAndGet(-countRelevantTransactions(event));
+    }
+
+    /**
+     * Notifies the controller of the next target consensus time.
+     *
+     * @param targetConsensusTime the next target consensus time
+     */
+    public void setNextTargetConsensusTime(@Nullable final Instant targetConsensusTime) {
```

Comment:
How do we decide the TCT ? Freeze tracking is easy, since it is stored in state. For scheduled transactions, I am assuming we need to have a background thread looking for next scheduled transaction time? is there a easy way to know the next scheduled transaction time? @tinker-michaelj

Reply:
If we have some threshold `T` when we want to quiesce the network, we can do something similar to [this](https://github.com/hiero-ledger/hiero-consensus-node/blob/b530d5ea10c154239cbb96e8ff85f4b18da55d23/hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/HandleWorkflow.java#L687-L687) on a background thread: 
```
var iter = scheduleService.executableTxns(
                    currentConsensusTime,
                    currentConsensusTime + T);
if (!iter.hasNext()) setNextTargetConsensusTime(currentConsensusTime + T);
``` 
Once we reach `currentConsensusTime + T`, we use `(currentConsensusTime + T) + T` in the same check